# Adam Neural Network in Task 1

### Given:
- Task 1 NN code
- α=0.001, β₁=0.9, β₂=0.999, ε=10⁻⁸.

### Task: 
- First Moment
- Second Moment
- Bias correction
- parameter update
- Compare vanilla nn with adam nn

## VANILLA NEURAL NETWORK

In [29]:
import numpy as np
#Parameters
Y = np.array([[1]])
X = np.array([[1],[2]])
W1 = np.array([[0.1,0.2],[-0.3,0.4],[0.5,-0.1]])
B1 = np.array([[0.1],[-0.2],[0.3]])
W2 = np.array([[0.3,-0.2,0.5]])
B2 = np.array([0.1])
alpha = 0.1
n = 1000
t = 100
#Activations

def sigmoid(z):
    return 1/(1+np.exp(-z))

def relu_grad(z):
    return (z>0).astype(float)

def relu(z):
    return np.maximum(0,z)

#Forward Pass

Z1 = np.dot(W1,X)+B1
A1 = relu(Z1)
Z2 = np.dot(W2,A1)+B2
A2 = sigmoid(Z2)
Loss = -np.log(A2)

print("y_hat = ",A2)
print("Loss = ",Loss)

#Backward Pass  
dZ2 = A2-Y
dW2 = np.dot(dZ2,A1.T)
dB2 = dZ2

dA1 = np.dot(W2.T,dZ2)
dZ1 = dA1*relu_grad(A1)
dW1 = np.dot(dZ1,X.T)
dB1 = dZ1

print("Initial Values")
print("Gradient wrt W1 = ",dW1)
print("Gradient wrt B1 = ",dB1)
print("Gradient wrt W2 = ",dW2)
print("Gradient wrt B2 = ",dB2)

#Gradient Descent (Single Iteration)
def gradient_descent():
    global W2,W1,B1,B2
    W2= W2-alpha*dW2
    W1= W1-alpha*dW1
    B1= B1-alpha*dB1
    B2= B2-alpha*dB2

#1000 gradient descent steps and printing after every 100 iterations
def forward_pass():
    global W2,Z1,A1,Z2,A2,X,W1,B1,B2, Loss
    Z1 = np.dot(W1,X)+B1
    A1 = relu(Z1)
    Z2 = np.dot(W2,A1)+B2
    A2 = sigmoid(Z2)
    Loss = -np.log(A2)

def backward_pass():
    global dZ2,dW2,dB2,dA1,dZ1,dW1,dB1
    dZ2 = A2-Y
    dW2 = np.dot(dZ2,A1.T)
    dB2 = dZ2

    dA1 = np.dot(W2.T,dZ2)
    dZ1 = dA1*relu_grad(A1)
    dW1 = np.dot(dZ1,X.T)
    dB1 = dZ1

for i in range(1,n+1):
    gradient_descent() #UPDATE W2 USING dW2
    forward_pass() #FIND Y_HAT AND LOSS
    backward_pass() #UPDATE dW2
    if(i%t==0):
        print(f"Iteration: {i}, Updated_W2 = {W2}, Updated y_hat = {A2}, Loss = {Loss}")

print("-----FINAL RESULTS AFTER 1000 ITERATIONS-----")
print(f"Final W2 = {W2}")
print(f"Final y_hat = {A2} ")


y_hat =  [[0.62714777]]
Loss =  [[0.46657309]]
Initial Values
Gradient wrt W1 =  [[-0.11185567 -0.22371134]
 [ 0.07457045  0.14914089]
 [-0.18642612 -0.37285223]]
Gradient wrt B1 =  [[-0.11185567]
 [ 0.07457045]
 [-0.18642612]]
Gradient wrt W2 =  [[-0.22371134 -0.11185567 -0.22371134]]
Gradient wrt B2 =  [[-0.37285223]]
Iteration: 100, Updated_W2 = [[ 0.81951496 -0.15487435  1.14917931]], Updated y_hat = [[0.99433331]], Loss = [[0.0056828]]
Iteration: 200, Updated_W2 = [[ 0.89219125 -0.15487435  1.24612665]], Updated y_hat = [[0.99771204]], Loss = [[0.00229058]]
Iteration: 300, Updated_W2 = [[ 0.93084542 -0.15487435  1.29784735]], Updated y_hat = [[0.99862913]], Loss = [[0.00137181]]
Iteration: 400, Updated_W2 = [[ 0.95687892 -0.15487435  1.33273351]], Updated y_hat = [[0.99904029]], Loss = [[0.00096017]]
Iteration: 500, Updated_W2 = [[ 0.97637594 -0.15487435  1.3588856 ]], Updated y_hat = [[0.99926967]], Loss = [[0.00073059]]
Iteration: 600, Updated_W2 = [[ 0.99189606 -0.15487435  1.3

## ADAM NEURAL NETWORK

In [7]:
import numpy as np
#Parameters

#α=0.001, β₁=0.9, β₂=0.999, ε=10⁻⁸.
Y = np.array([[1]])
X = np.array([[1],[2]])
W1 = np.array([[0.1,0.2],[-0.3,0.4],[0.5,-0.1]])
B1 = np.array([[0.1],[-0.2],[0.3]])
W2 = np.array([[0.3,-0.2,0.5]])
B2 = np.array([0.1])

alpha = 0.001
beta1 = 0.9
beta2 = 0.999
epsilon = 10**-8
n = 1000
t = 100

#Activations

def sigmoid(z):
    return 1/(1+np.exp(-z))

def relu_grad(z):
    return (z>0).astype(float)

def relu(z):
    return np.maximum(0,z)

#Forward Pass

Z1 = np.dot(W1,X)+B1
A1 = relu(Z1)
Z2 = np.dot(W2,A1)+B2
A2 = sigmoid(Z2)
Loss = -np.log(A2)

print("y_hat = ",A2)
print("Loss = ",Loss)

#Backward Pass  
dZ2 = A2-Y
dW2 = np.dot(dZ2,A1.T)
dB2 = dZ2

dA1 = np.dot(W2.T,dZ2)
dZ1 = dA1*relu_grad(A1)
dW1 = np.dot(dZ1,X.T)
dB1 = dZ1

print("Initial Values")
print("Gradient wrt W1 = ",dW1)
print("Gradient wrt B1 = ",dB1)
print("Gradient wrt W2 = ",dW2)
print("Gradient wrt B2 = ",dB2)

#Variables for Adam
Mt_W2 = np.zeros(np.shape(dW2))
Mt_W1 = np.zeros(np.shape(dW1))
Mt_B2 = np.zeros(np.shape(dB2))
Mt_B1 = np.zeros(np.shape(dB1))

unbiased_Mt_W2 = np.zeros(np.shape(dW2))
unbiased_Mt_W1 = np.zeros(np.shape(dW1))
unbiased_Mt_B2 = np.zeros(np.shape(dB2))
unbiased_Mt_B1 = np.zeros(np.shape(dB1))

Vt_W2 = np.zeros(np.shape(dW2))
Vt_W1 = np.zeros(np.shape(dW1))
Vt_B2 = np.zeros(np.shape(dB2))
Vt_B1 = np.zeros(np.shape(dB1))

unbiased_Vt_W2 = np.zeros(np.shape(dW2))
unbiased_Vt_W1 = np.zeros(np.shape(dW1))
unbiased_Vt_B2 = np.zeros(np.shape(dB2))
unbiased_Vt_B1 = np.zeros(np.shape(dB1))

lr_W2 = np.zeros(np.shape(dW2))
lr_W1 = np.zeros(np.shape(dW1))
lr_B2 = np.zeros(np.shape(dB2))
lr_B1 = np.zeros(np.shape(dB1))

#### This is where things change in Adam
def update_mt(t):
    global dB2, dW2, dB1, dB2, unbiased_Mt_W2, unbiased_Mt_W1, unbiased_Mt_B2, unbiased_Mt_B1, Mt_B1, Mt_W2, Mt_B2, Mt_W1
    Mt_W2 = beta1*Mt_W2  + (1-beta1)*(dW2)
    Mt_W1 = beta1*Mt_W1  + (1-beta1)*(dW1)
    Mt_B1 = beta1*Mt_B1 + (1-beta1)*(dB1)
    Mt_B2= beta1*Mt_B2 + (1-beta1)*(dB2)
    #unbias
    unbiased_Mt_W2 = Mt_W2/(1-beta1**t)
    unbiased_Mt_W1 = Mt_W1/(1-beta1**t)
    unbiased_Mt_B2 = Mt_B2/(1-beta1**t)
    unbiased_Mt_B1 = Mt_B1/(1-beta1**t)


def update_vt(t):
    global beta1, beta2, dB2, dW2, dB1, dB2, unbiased_Vt_B1, unbiased_Vt_W1, unbiased_Vt_B2, unbiased_Vt_W2, Vt_B1, Vt_B2, Vt_W1, Vt_W2
    Vt_W2 = beta2*Vt_W2  + (1-beta2)*(dW2**2)
    Vt_W1 = beta2*Vt_W1  + (1-beta2)*(dW1**2)
    Vt_B1 = beta2*Vt_B1 + (1-beta2)*(dB1**2)
    Vt_B2= beta2*Vt_B2 + (1-beta2)*(dB2**2)
    #unbias 
    unbiased_Vt_W2 = Vt_W2/(1-beta2**t)
    unbiased_Vt_W1 = Vt_W1/(1-beta2**t)
    unbiased_Vt_B2 = Vt_B2/(1-beta2**t)
    unbiased_Vt_B1 = Vt_B1/(1-beta2**t)

def update_lr():
    global W2,W1,B1,B2,alpha, unbiased_Vt_W2, epsilon, lr_B1, lr_W2, lr_W1, lr_B2, unbiased_Vt_W1, unbiased_Vt_B1, unbiased_Vt_B2
    lr_W2 = (alpha)/((unbiased_Vt_W2+epsilon)**0.5)
    lr_W1 = (alpha)/((unbiased_Vt_W1+epsilon)**0.5)
    lr_B1 = (alpha)/((unbiased_Vt_B1+epsilon)**0.5)
    lr_B2 = (alpha)/((unbiased_Vt_B2+epsilon)**0.5)

#Gradient Descent (Single Iteration)
def gradient_descent(t):
    global W2,W1,B1,B2,lr_W2,lr_W1,lr_B2, lr_B1, unbiased_Mt
    update_mt(t)
    update_vt(t)
    update_lr()
    W2= W2-lr_W2*unbiased_Mt_W2
    W1= W1-lr_W1*unbiased_Mt_W1
    B1= B1-lr_B1*unbiased_Mt_B1
    B2= B2-lr_B2*unbiased_Mt_B2
    

#1000 gradient descent steps and printing after every 100 iterations
def forward_pass():
    global W2,Z1,A1,Z2,A2,X,W1,B1,B2, Loss
    Z1 = np.dot(W1,X)+B1
    A1 = relu(Z1)
    Z2 = np.dot(W2,A1)+B2
    A2 = sigmoid(Z2)
    Loss = -np.log(A2)

def backward_pass():
    global dZ2,dW2,dB2,dA1,dZ1,dW1,dB1
    dZ2 = A2-Y
    dW2 = np.dot(dZ2,A1.T)
    dB2 = dZ2

    dA1 = np.dot(W2.T,dZ2)
    dZ1 = dA1*relu_grad(A1)
    dW1 = np.dot(dZ1,X.T)
    dB1 = dZ1

for i in range(1,n+1):
    gradient_descent(i) #UPDATE W2 USING dW2
    forward_pass() #FIND Y_HAT AND LOSS
    backward_pass() #UPDATE dW2
    if(i%t==0):
        print(f"Iteration: {i}, Updated_W2 = {W2}, Updated y_hat = {A2}, Loss = {Loss}")

print("-----FINAL RESULTS AFTER 1000 ITERATIONS-----")
print(f"Final W2 = {W2}")
print(f"Final y_hat = {A2} ")


y_hat =  [[0.62714777]]
Loss =  [[0.46657309]]
Initial Values
Gradient wrt W1 =  [[-0.11185567 -0.22371134]
 [ 0.07457045  0.14914089]
 [-0.18642612 -0.37285223]]
Gradient wrt B1 =  [[-0.11185567]
 [ 0.07457045]
 [-0.18642612]]
Gradient wrt W2 =  [[-0.22371134 -0.11185567 -0.22371134]]
Gradient wrt B2 =  [[-0.37285223]]
Iteration: 100, Updated_W2 = [[ 0.40102416 -0.14201015  0.60091325]], Updated y_hat = [[0.76476171]], Loss = [[0.26819098]]
Iteration: 200, Updated_W2 = [[ 0.49441081 -0.14168061  0.69357687]], Updated y_hat = [[0.86077293]], Loss = [[0.14992453]]
Iteration: 300, Updated_W2 = [[ 0.57062697 -0.1416806   0.76858279]], Updated y_hat = [[0.91790597]], Loss = [[0.08566032]]
Iteration: 400, Updated_W2 = [[ 0.63047144 -0.1416806   0.82704793]], Updated y_hat = [[0.94893405]], Loss = [[0.05241598]]
Iteration: 500, Updated_W2 = [[ 0.6779679  -0.1416806   0.87316568]], Updated y_hat = [[0.96616476]], Loss = [[0.0344209]]
Iteration: 600, Updated_W2 = [[ 0.71661607 -0.1416806   0.9

y_hat =  [[0.62714777]]
Loss =  [[0.46657309]]
Initial Values
Gradient wrt W1 =  [[-0.11185567 -0.22371134]
 [ 0.07457045  0.14914089]
 [-0.18642612 -0.37285223]]
Gradient wrt B1 =  [[-0.11185567]
 [ 0.07457045]
 [-0.18642612]]
Gradient wrt W2 =  [[-0.22371134 -0.11185567 -0.22371134]]
Gradient wrt B2 =  [[-0.37285223]]
Iteration: 100, Updated_W2 = [[ 0.81951496 -0.15487435  1.14917931]], Updated y_hat = [[0.99433331]], Loss = [[0.0056828]]
Iteration: 200, Updated_W2 = [[ 0.89219125 -0.15487435  1.24612665]], Updated y_hat = [[0.99771204]], Loss = [[0.00229058]]
Iteration: 300, Updated_W2 = [[ 0.93084542 -0.15487435  1.29784735]], Updated y_hat = [[0.99862913]], Loss = [[0.00137181]]
Iteration: 400, Updated_W2 = [[ 0.95687892 -0.15487435  1.33273351]], Updated y_hat = [[0.99904029]], Loss = [[0.00096017]]
Iteration: 500, Updated_W2 = [[ 0.97637594 -0.15487435  1.3588856 ]], Updated y_hat = [[0.99926967]], Loss = [[0.00073059]]
Iteration: 600, Updated_W2 = [[ 0.99189606 -0.15487435  1.37971774]], Updated y_hat = [[0.99941458]], Loss = [[0.0005856]]
Iteration: 700, Updated_W2 = [[ 1.00475103 -0.15487435  1.39698171]], Updated y_hat = [[0.99951378]], Loss = [[0.00048634]]
Iteration: 800, Updated_W2 = [[ 1.01570026 -0.15487435  1.41169263]], Updated y_hat = [[0.99958565]], Loss = [[0.00041444]]
Iteration: 900, Updated_W2 = [[ 1.02522172 -0.15487435  1.42448982]], Updated y_hat = [[0.99963993]], Loss = [[0.00036014]]
Iteration: 1000, Updated_W2 = [[ 1.03363518 -0.15487435  1.43580122]], Updated y_hat = [[0.99968227]], Loss = [[0.00031778]]
-----FINAL RESULTS AFTER 1000 ITERATIONS-----
Final W2 = [[ 1.03363518 -0.15487435  1.43580122]]
Final y_hat = [[0.99968227]]